File: src/main/kotlin/models/RussianAddressBuilder.kt
package org.example.models

import org.example.enums.AddressObjectType
import org.example.enums.InhabitedLocalityType
import org.example.enums.MunicipalityType
import org.example.enums.ParkingSpaceType
import org.example.enums.PlanningStructureType
import org.example.enums.PremiseType
import org.example.enums.RoadNetworkType
import org.example.enums.SettlementType
import org.example.enums.SubjectRfType

class RussianAddressBuilder: IBuilder {
    private val address = RussianAddress()

    fun addSubject(type: SubjectRfType, name: String) {
        address.subject = SubjectInfo(type, name)
    }

    fun addMunicipality(type: MunicipalityType, name: String) {
        address.municipality = MunicipalityInfo(type, name)
    }

    fun addSettlement(type: SettlementType, name: String) {
        address.settlement = SettlementInfo(type, name)
    }

    fun addLocality(type: InhabitedLocalityType, name: String) {
        address.locality = LocalityInfo(type, name)
    }

    fun addPlanningStructure(type: PlanningStructureType, name: String) {
        address.planningStructure = PlanningStructureInfo(type, name)
    }

    fun addRoadNetwork(type: RoadNetworkType, name: String) {
        address.roadNetwork = RoadNetworkInfo(type, name)
    }

    fun addLandPlot(number: String) {
        address.addressObjectType = AddressObjectType.LandPlot
        address.landPlotNumber = number
        // Сброс несовместимых полей
        address.buildingNumber = null
        address.buildingBlock = null
        address.buildingStructure = null
        address.buildingEstate = null
        address.premiseType = null
        address.premiseNumber = null
        address.parkingSpaceType = null
        address.parkingSpaceNumber = null
    }

    fun addBuilding(number: String, block: String? = null, structure: String? = null, estate: String? = null) {
        address.addressObjectType = AddressObjectType.Building  // или Structure/Construction, но по умолчанию Building
        address.buildingNumber = number
        address.buildingBlock = block
        address.buildingStructure = structure
        address.buildingEstate = estate
        // Сброс несовместимых полей
        address.landPlotNumber = null
        address.premiseType = null
        address.premiseNumber = null
        address.parkingSpaceType = null
        address.parkingSpaceNumber = null
    }

    fun addPremise(type: PremiseType, number: String) {
        require(address.parkingSpaceType == null) { "Нельзя добавить помещение, если уже добавлено машино-место" }
        if (address.addressObjectType == null) {
            address.addressObjectType = AddressObjectType.Building
        }
        address.premiseType = type
        address.premiseNumber = number
    }

    fun addParkingSpace(type: ParkingSpaceType, number: String) {
        require(address.premiseType == null) { "Нельзя добавить машино-место, если уже добавлено помещение" }
        if (address.addressObjectType == null) {
            address.addressObjectType = AddressObjectType.Building
        }
        address.parkingSpaceType = type
        address.parkingSpaceNumber = number
    }

    fun addPostalCode(postalCode: String) {
        address.postalCode = postalCode
    }

    override fun build(): RussianAddress {
        // Валидация перед возвратом копии
        when (address.addressObjectType) {
            AddressObjectType.LandPlot -> {
                require(!address.landPlotNumber.isNullOrBlank()) { "Для земельного участка должен быть указан номер" }
                require(address.buildingNumber == null && address.premiseType == null && address.parkingSpaceType == null) {
                    "Для земельного участка не могут быть указаны buildingNumber, premiseType или parkingSpaceType"
                }
            }
            AddressObjectType.Building, AddressObjectType.Structure, AddressObjectType.Construction -> {
                require(!address.buildingNumber.isNullOrBlank()) { "Для здания/строения/сооружения должен быть указан номер" }
                require(address.landPlotNumber == null) { "Для здания/строения/сооружения не может быть указан landPlotNumber" }
            }
            null -> { /* нет дополнительных требований */ }
        }
        require(!(address.premiseType != null && address.parkingSpaceType != null)) {
            "Нельзя одновременно указать помещение и машино-место"
        }

        return address
    }
}

File: Csharp/RussianAddress/Models/RussianAddressBuilder.cs
using System;
using RussianAddress.Enums;

namespace RussianAddress.Models;

public class RussianAddressBuilder
{
    private readonly RussianAddress _address;
    
    public RussianAddressBuilder(SubjectRfType subjectType, string subjectName)
    {
        var subjectInfo = new SubjectInfo { Type = subjectType, Name = subjectName };
        var address = new RussianAddress() { Subject =  subjectInfo };
        _address = address;
    }

    public void AddSubject(SubjectRfType type, string name)
    {
        _address.Subject = new SubjectInfo { Type = type, Name = name };
    }

    public void AddMunicipality(MunicipalityType type, string name)
    {
        _address.Municipality = new MunicipalityInfo { Type = type, Name = name };
    }

    public void AddSettlement(SettlementType type, string name)
    {
        _address.Settlement = new SettlementInfo { Type = type, Name = name };
    }

    public void AddLocality(InhabitedLocalityType type, string name)
    {
        _address.Locality = new LocalityInfo { Type = type, Name = name };
    }

    public void AddPlanningStructure(PlanningStructureType type, string name)
    {
        _address.PlanningStructure = new PlanningStructureInfo { Type = type, Name = name };
    }

    public void AddRoadNetwork(RoadNetworkType type, string name)
    {
        _address.RoadNetwork = new RoadNetworkInfo { Type = type, Name = name };
    }

    public void AddLandPlot(string number)
    {
        // При добавлении земельного участка сбрасываем все, что связано со зданием, помещением и машино-местом
        _address.AddressObjectType = AddressObjectType.LandPlot;
        _address.LandPlotNumber = number;

        _address.BuildingNumber = null;
        _address.BuildingBlock = null;
        _address.BuildingStructure = null;
        _address.BuildingEstate = null;
        _address.PremiseType = null;
        _address.PremiseNumber = null;
        _address.ParkingSpaceType = null;
        _address.ParkingSpaceNumber = null;
    }

    public void AddBuilding(string number, AddressObjectType type, string? block = null, string? structure = null, string? estate = null)
    {
        _address.AddressObjectType = type;
        _address.BuildingNumber = number;
        _address.BuildingBlock = block;
        _address.BuildingStructure = structure;
        _address.BuildingEstate = estate;

        _address.LandPlotNumber = null;
        _address.PremiseType = null;
        _address.PremiseNumber = null;
        _address.ParkingSpaceType = null;
        _address.ParkingSpaceNumber = null;
    }

    public void AddPremise(PremiseType type, string number)
    {
        if (_address.ParkingSpaceType != null)
            throw new InvalidOperationException("Нельзя добавить помещение, если уже добавлено машино-место.");
        _address.AddressObjectType ??= AddressObjectType.Building;

        _address.PremiseType = type;
        _address.PremiseNumber = number;
    }

    public void AddParkingSpace(ParkingSpaceType type, string number)
    {
        if (_address.PremiseType != null)
            throw new InvalidOperationException("Нельзя добавить машино-место, если уже добавлено помещение.");
        _address.AddressObjectType ??= AddressObjectType.Building;

        _address.ParkingSpaceType = type;
        _address.ParkingSpaceNumber = number;
    }

    public void AddPostalCode(string postalCode)
    {
        _address.PostalCode = postalCode;
    }

    public RussianAddress Build()
    {
        // Дополнительные проверки корректности (опционально)
        if (_address.AddressObjectType == AddressObjectType.LandPlot && string.IsNullOrWhiteSpace(_address.LandPlotNumber))
            throw new InvalidOperationException("Для земельного участка должен быть указан номер.");
        if (_address.AddressObjectType != AddressObjectType.LandPlot && _address.LandPlotNumber != null)
            throw new InvalidOperationException("Номер земельного участка может быть указан только для типа LandPlot.");

        if ((_address.AddressObjectType is AddressObjectType.Building or AddressObjectType.Structure ||
             _address.AddressObjectType == AddressObjectType.Construction) &&
            string.IsNullOrWhiteSpace(_address.BuildingNumber))
            throw new InvalidOperationException("Для здания/строения/сооружения должен быть указан номер.");

        if (_address.PremiseType != null && _address.ParkingSpaceType != null)
            throw new InvalidOperationException("Нельзя одновременно указать помещение и машино-место.");
        
        return _address;
    }
}